packages/atclient/src/atclient.c

```diff
@@ -186,19 +186,30 @@ bool atclient_is_atsign_initialized(const atclient *ctx) {
   return ctx->_initialized_fields[ATCLIENT_ATSIGN_INDEX] & ATCLIENT_ATSIGN_INITIALIZED;
 }
 
-int atclient_pkam_authenticate(atclient *ctx, const char *atserver_host, const int atserver_port,
-                               const atclient_atkeys *atkeys, const char *atsign) {
+int atclient_pkam_authenticate(atclient *ctx, const atclient_atkeys *atkeys, const char *atsign, atclient_pkam_authenticate_options *options) {
 
   int ret = 1; // error by default
+  if(options == NULL){
+    atclient_pkam_authenticate_options options;
+    atclient_pkam_authenticate_options_init(&options);  
+  }
 
   /*
    * 1. Validate arguments
    */
-  if ((ret = atclient_pkam_authenticate_validate_arguments(ctx, atserver_host, atserver_port, atkeys, atsign)) != 0) {
+  if ((ret = atclient_pkam_authenticate_validate_arguments(ctx, atkeys, atsign)) != 0) {
     atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "atclient_pkam_authenticate_validate_arguments: %d\n", ret);
     return ret;
   }
 
+  /*
+  * 1.1 If atserver_host and atserver_port are null and 0 respectively, then fetch the server host and port.
+  */
+  if(options->atserver_host == NULL && options->atserver_port == 0){
+    atclient_utils_find_atserver_address(options->root_server_host, options->root_server_port, atsign, &(options->atserver_host), &(options->atserver_port));
```

Comment:
This function `atclient_utils_find_atserver_address` returns an error code which should be handled.